orm/model_bucket.go

```diff
@@ -0,0 +1,174 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// TODO
+// - migrations
+// - do not use Bucket but directly access KVStore
+// - register for queries
+
+// Model is impelemented by any entity that can be stored using ModelBucket.
+//
+// This is the same interface as CloneableData. Using the right type names
+// provides an easier to read API.
+type Model interface {
+	weave.Persistent
+	Validate() error
+	Copy() CloneableData
+}
+
+// ModelSlicePtr represents a pointer to a slice of models. Think of it as
+// *[]Model Because of Go type system, using []Model would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type ModelSlicePtr interface{}
+
+// ModelBucket is implemented by buckets that operates on Models rather than
+// Objects.
+type ModelBucket interface {
+	// One query the database for a single model instance. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// model.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given model type cannot be used to contain stored entity, ErrType
+	// is returned.
+	One(db weave.ReadOnlyKVStore, key []byte, dest Model) error
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is retured and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest ModelSlicePtr) error
+
+	// Put saves given model in the database. Before inserting into
+	// database, model is validated using its Validate method.
+	// If the key is nil or zero length then a sequence generator is used
+	// to create a unique key value.
+	// Using a key that alreayd exists in the database cause the value to
+	// be overwritten.
+	Put(db weave.KVStore, key []byte, m Model) ([]byte, error)
+
+	// Delete removes an entity with given primary key from the database.
+	// It returns ErrNotFound if an entity with given key does not exist.
+	Delete(db weave.KVStore, key []byte) error
+
+	// Register registers this buckets content to be accessable via query
+	// requests under the given name.
+	Register(name string, r weave.QueryRouter)
+}
+
+// NewModelBucket returns a ModelBucket instance. This implementation relies on
+// a bucket instance. Final implementation should operate directly on the
+// KVStore instead.
+func NewModelBucket(b Bucket) ModelBucket {
+	return &modelBucket{
+		b:     b,
+		idSeq: b.Sequence("id"),
+	}
+}
+
+type modelBucket struct {
```

Comment:
We could add here the expected model type to be used (either struct or ptr to struct).
Then auto-enforce that in all Puts.

That is a bunch of the boiler-plate in `Save()` to ensure they write the proper data format. Since we use reflect heavily here anyway, let's just auto-magically enforce proper type on writes (Buckets must all be the same type)

Reply:
Yes! That would produce an ugly API though. Instead of [the current](https://github.com/iov-one/weave/pull/615/files/e5668dfb181bbca7436fa6c7648cc29fb83d67a6#diff-7c3e875063ade4e83ecfc8e98b96235aR164):
```go
b := orm.NewModelBucket(
    orm.NewBucket("mymodel", orm.NewSimpleObj(nil, &MyModel{})),
)
```
the API would require to pass the `MyModel` reference twice and be 
```go
b := orm.NewModelBucket(
    &MyModel{},
    orm.NewBucket("mymodel", orm.NewSimpleObj(nil, &MyModel{})),
)
```
I believe this will be possible and look good once model bucket will not rely on the original bucket implementation.


This was addressed in https://github.com/iov-one/weave/pull/615#discussion_r284143700